### PR TITLE
8255397: x86: coalesce reference and int entry points into vtos bytecodes

### PR DESCRIPTION
**Pull request by cl4es**
On x86 - both 32- and 64-bit - the code laid out for transitionining into a vtos bytecode when having a reference and int top-of-stack state is semantically identical, and can be coalesced.

This patch removes a short jump on some cases which is marginally beneficial when interpreting, while measurably reducing overhead of generating the interpreter itself.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ⏳ (7/9 running) | ⏳ (3/9 running) |

### Issue
 * [JDK-8255397](https://bugs.openjdk.java.net/browse/JDK-8255397): x86: coalesce reference and int entry points into vtos bytecodes

### Reviewers
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**)
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/865/head:pull/865`
`$ git checkout pull/865`

### CODE REVIEW

**Review by shipilev — COMMENTED** [review]
It rubs me the wrong way that we are effectively changing `push_ptr` to `push_i` for `aep`. While it is implemented in the same manner in `interp_masm_x86.cpp` -- delegating to `push`, it still means if `push_i` implementation changes, `aep` would do the `push_i` _as if_ it is integer, not pointer. Ditto a change in `push_ptr` (adding verification, maybe?) would miss this code.

So, how much of the improvement we are talking about to sacrifice this?

**Comment by coleenp** [comment]
Yes, I had the same queasy feeling looking at this.  It's just unclear at this callsite that push_i doesn't do the wrong thing for 64 bit.

**Comment by cl4es** [comment]
> It rubs me the wrong way that we are effectively changing `push_ptr` to `push_i` for `aep`. While it is implemented in the same manner in `interp_masm_x86.cpp` -- delegating to `push`, it still means if `push_i` implementation changes, `aep` would do the `push_i` _as if_ it is integer, not pointer. Ditto a change in `push_ptr` (adding verification, maybe?) would miss this code.

Verification is done explicitly with `__ verify_oop(..)` and friends, so it seems unlikely we'll overload `push_ptr` any time soon (and they have been semantically identical for many years, even before the merging of 32- and 64-bit `interp_masm_x86...`). But I acknowledge this adds a fragility here, but perhaps there are some assertions we can add to put a check that `push_ptr` and `push_i` stays semantically the same?

> 
> So, how much of the improvement we are talking about to sacrifice this?

A few hundred thousand instructions and branches on Hello World (seems unconditional jumps are logged as branches by `perf`?):

Baseline:
```
       103,795,433      instructions              #    0.59  insn per cycle           ( +-  0.07% )
        20,263,519      branches                  #  200.867 M/sec                    ( +-  0.08% )
           731,187      branch-misses             #    3.61% of all branches          ( +-  0.15% )       0.067306367 seconds time elapsed                                          ( +-  0.24% )
```

Patch:
```
       103,466,523      instructions              #    0.59  insn per cycle           ( +-  0.07% )
        20,068,162      branches                  #  201.935 M/sec                    ( +-  0.08% )
           727,575      branch-misses             #    3.63% of all branches          ( +-  0.13% )       0.066568115 seconds time elapsed                                          ( +-  0.27% )
```

For Hello World maybe half of that comes from reduced overhead of generating, the rest from quickening quite a few bytecode transitions. There's a scaling component (seen a few million instruction gains on slightly larger apps), but it's nothing huge.

**Comment by shipilev** [comment]
Okay, so that is 0.3% less instructions and ~1% less branches on Hello World. That's interesting. 

Would rebalancing the entry points order give the similar improvement without messing up the code? For example, what happens if we move `aep` to be the last entry point, and set up `[bcsi]ep` for a short jump?

There is a middle-ground, I think: introduce `push_i_or_ptr` and delegate it to `push`. That would make it clear what usages expect `push_i` and `push_ptr` shapes to match, and if later it proves to be a problem, we could easily revert all new usages to the old form.

File: src/hotspot/cpu/x86/interp_masm_x86.cpp [hunk 1]
```diff
@@ -605,6 +605,10 @@ void InterpreterMacroAssembler::push_i(Register r) {
   push(r);
 }
 
+void InterpreterMacroAssembler::push_i_or_ptr(Register r) {
+  push_i(r);
```
Comment (shipilev) [inline src/hotspot/cpu/x86/interp_masm_x86.cpp:609]: Should be `push(r)`: it is both cleaner and avoids a middle call to `push_i(r)`.

File: src/hotspot/cpu/x86/interp_masm_x86.hpp [hunk 2]
```diff
@@ -139,9 +139,19 @@ class InterpreterMacroAssembler: public MacroAssembler {
...
+  // On x86, pushing a ptr or an int is semantically identical, but we
+  // maintain a distinction for clarity and for making it easier to change
+  // semantics in the future
   void push_ptr(Register r = rax);
   void push_i(Register r = rax);
 
+  // push_i_or_ptr is provided for when explicitly allowing either a ptr or
+  // an int might have some advantage, while still documenting the fact that a
+  // ptr might be pushed to the stack. This method will never do any
+  // verification of the oop.
```
Comment (shipilev) [inline src/hotspot/cpu/x86/interp_masm_x86.hpp:152]: I don't think we need to mention verification here.

**Review by shipilev — CHANGES_REQUESTED** [review]

File: src/hotspot/cpu/x86/interp_masm_x86.hpp (hunk 2 above)
Comment (cl4es) [reply src/hotspot/cpu/x86/interp_masm_x86.hpp:152]: Sure, I'll drop that.

**Comment by cl4es** [comment]
> Would rebalancing the entry points order give the similar improvement without messing up the code? For example, what happens if we move `aep` to be the last entry point, and set up `[bcsi]ep` for a short jump?

Looks like a wash, at least on Hello World

> There is a middle-ground, I think: introduce `push_i_or_ptr` and delegate it to `push`. That would make it clear what usages expect `push_i` and `push_ptr` shapes to match, and if later it proves to be a problem, we could easily revert all new usages to the old form.

Good suggestion.

**Review by shipilev — APPROVED** [review]
This looks good to me. Coleen needs to ack too.

**Review by coleenp — APPROVED** [review]
Yes that was a good suggestion @shipilev and the change looks much better to me now.  Thanks!

**Comment by cl4es** [comment]
@shipilev @coleenp - thank you for reviewing!

/integrate

**Pull request closed without merging**